src/utils/BucketUtils.ts

```diff
@@ -0,0 +1,384 @@
+  return dar.elections
+    ? Object.values(dar.elections)
+        .filter((e: Election) => datasetIds.includes(e.datasetId))
+    : []
+}
+
+/**
+ * Optionally filter a list of collection datasets by the dac ids provided.
+ * @private
+ * @param dacIds List of dac ids. Can be empty
+ * @param datasets List of datasets to filter
+ * @returns {Dataset[]}
+ */
+const filterDatasetsByDACs = (dacIds: number[], datasets: Dataset[]): Dataset[] => {
+  return isEmpty(dacIds)
+    ? datasets
+    : filter(
+        (dataset: Dataset) => includes(dataset.dacId)(dacIds),
+      )(datasets)
+}
+
+/**
+ * Generate the summary of algorithm results suitable for display in the UI
+ *
+ * Four potential cases:
+ *  1. No matches
+ *  2. Exactly one match or N matches that are all the same - easy case
+ *  3. Abstain is true and match is false - decision is abstained
+ *  4. N matches - not all the same - very confusing case
+ *
+ * @private
+ * @param bucket
+ * @returns {AlgorithmResult}
+ */
+const calculateAlgorithmResultForBucket = (bucket: Bucket): AlgorithmResult => {
+  // V1 and V2: We actually DO NOT want to show system match results when the data use indicates
+  // that a match should not be made. This happens for all "Other" cases.
+  const algorithmVersionV3 = bucket.matchResults.length > 0 && bucket.matchResults[0].algorithmVersion === 'v3'
+  const unmatchable = isOther(bucket.dataUse) || shouldAbstain(bucket.dataUse)
+  // Check on all possible true/false values in the matches.
+  // If all matches are the same, we can merge them into a single match object for display.
+  // If they are not all the same, we have to punt this decision solely to the DAC.
+  // Check algorithm version: V3 does not need to be checked for 'unmatchable'
+  const matchVals: boolean[] = (algorithmVersionV3 || !unmatchable)
+    ? flow(
+        map((m: MatchResult) => m.match),
+        uniq,
+      )(bucket.matchResults)
+    : []
+
+  const abstain = processV3Abstain(bucket.matchResults)
+
+  // check results based on matchVals
+  if (isEmpty(matchVals)) {
+    return { result: 'N/A', createDate: undefined, rationales: undefined, id: bucket.key }
+  }
+  else if ((matchVals.length === 1)) {
+    const rationales: string[] = flow(
+      flatMap((match: MatchResult) => match.rationales),
+      uniq,
+    )(bucket.matchResults)
+    const { createDate, failed, id, match } = bucket.matchResults[0]
+    const matchResult = { createDate, rationales, failed, id, match }
+    if (abstain) {
+      return {
+        result: 'Abstain',
+        createDate,
+        rationales,
+        id,
+      }
+    }
+    else {
+      return {
+        result: processMatchData(matchResult),
+        createDate,
+        rationales,
+        id,
+      }
+    }
+  }
+  else {
+    // Different match values? Provide a custom message
+    return {
+      result: 'Unable to determine a system match',
```

Comment:
Do we define "system match" in prose anywhere for users?  This phrase is probably worth using in forthcoming RADAR docs, at least.

Reply:
Great point. `system match`, in this context, is the Ontology algorithm match described [here](https://github.com/DataBiosphere/consent-ontology/blob/develop/docs/Algorithms.md). We definitely could use some better instructional language in the UI around this point.